### PR TITLE
route /api/topics page and create tests 

### PR DESCRIPTION
**Pull request by jimmybombimmy**
api page created and tests created

### CODE REVIEW

File: __tests__/api.test.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+const request = require('supertest');
+
+const seed = require('../db/seeds/seed.js')
+const testData = require('../db/data/test-data')
+const connection = require('../db/connection.js')
+
+beforeEach(() => seed(testData))
+afterAll(() => connection.end());
+
+describe.only("GET /api/topics", () => {
```
Comment (Keb84) [inline __tests__/api.test.js]: Please remove your .only's before sending any pull requests 👀

File: __tests__/api.test.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      .expect(200)
+      .then(({body}) => {
+        expect(body).toHaveLength(3)
+      })
+  })
+  test("404: Come back with error if route not correct", () => {
+    return request(app)
+      .get("/api/tops")
+      .expect(404)
+  })
```
Comment (Keb84) [inline __tests__/api.test.js:35]: This 404 test is not specific to this endpoint and so should be in its own describe block 🔎

File: __tests__/api.test.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        expect(body[0]).toHaveProperty("description", expect.any(String))
+      })
+  })
+  test("200: Check that all of the columns have the right length", () => {
+    return request(app)
+      .get("/api/topics")
+      .expect(200)
+      .then(({body}) => {
+        expect(body).toHaveLength(3)
+      })
```
Comment (Keb84) [inline __tests__/api.test.js]: These 3 test blocks could all be condensed to 1 and you could just have the expect statements after each other 💭

File: __tests__/api.test.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  })
+  test("200: Check that all of the columns match their respective names", () => {
+    return request(app)
+      .get("/api/topics")
+      .expect(200)
+      .then(({body}) => {
+        expect(body[0]).toHaveProperty("slug", expect.any(String))
+        expect(body[0]).toHaveProperty("description", expect.any(String))
+      })
```
Comment (Keb84) [inline __tests__/api.test.js]: As you know the length of the array is 3 would this be a more robust test if you checked each of them to ensure they all have the correct keys 🤔

File: app.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+} = require("./db/controllers/data.controller.js")
+
+const app = express();
+
+app.get("/api/topics", topicsRoute);
+
+app.listen(9090, (err) => {
+  if (err) console.log("Error running on server")
+  else console.log("Server running on port 9090")
+})
```
Comment (Keb84) [inline app.js]: Should your app.listen live in your app? Think about extracting this out into it's own listen.js file 🚀

File: app.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
+const express = require("express");
+const fs = require("fs/promises");
```
Comment (Keb84) [inline app.js]: Do you really need this here 🤔

File: db/controllers/data.controller.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
+
+const pool = require('../connection.js')
+const queries = require('../queries/queries.js')
+
+exports.topicsRoute = (req, res) => {
+  pool.query(queries.getTopics, (err, results) => {
```
Comment (Keb84) [inline db/controllers/data.controller.js]: We've been teaching the MVC structure so we'd like you to follow it here. Please extract this into a model and use your model to make queries to the database.

File: db/controllers/data.controller.js [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+    if (err) throw err;
+    res.status(200).json(results.rows)
+  })
+
```
Comment (Keb84) [inline db/controllers/data.controller.js]: A catch block would be good here rather than throwing your error so that it will use the middleware 🤸

**Review by Keb84 — CHANGES_REQUESTED** [review]
Nice work on ticket 2. Please check out my comments, make the changes and resubmit 🚀

File: __tests__/api.test.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        .expect(200)
+        .then(({
+          body
+        }) => {
+          expect(body).toMatchObject(testData.topicData)
+          expect(body).toHaveLength(3)
+          for (let i = 0; i < body.length; i++) {
+            expect(body[0]).toHaveProperty("slug", expect.any(String))
+            expect(body[0]).toHaveProperty("description", expect.any(String))
+          }
```
Comment (Keb84) [inline __tests__/api.test.js:25]: As you are only checking the first index in your expect statements is this loop doing anything? 👀 . Now you have a length check would a forEach maybe more appropriate here? You could then check all the indexes have the properties you expect. 

Nice work of the use expect.any() for your values making them nice and dynamic 🤸

File: __tests__/api.test.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+describe("GET /api/topics", () => {
+  describe("successful connection test(s)", () => {
+    test("200: expect page to respond with corresponding topics table", () => {
+      return request(app)
+        .get("/api/topics")
+        .expect(200)
+        .then(({
+          body
+        }) => {
+          expect(body).toMatchObject(testData.topicData)
```
Comment (Keb84) [inline __tests__/api.test.js]: Using this expect statement here for topics is ok but with wouldn't be a robust test for articles or comments as they will have an id once inserted into the database.  toMatchObject may also not be the most appropriate matcher here 🤔

File: app.js [hunk 11]
```diff
@@ -0,0 +1,15 @@
+const express = require("express");
+const serverInfo = require("./db/server.js")
+const {
+  getTopics
+} = require("./db/controllers/data.controller.js")
+
+const app = express();
+
+app.set('view engine', 'ejs')
```
Comment (Keb84) [inline app.js]: Do you know what this is doing here? Is this needed for this ticket? Please remove this line of code 👀

File: app.js [hunk 12]
```diff
@@ -0,0 +1,15 @@
+const express = require("express");
+const serverInfo = require("./db/server.js")
```
Comment (Keb84) [inline app.js]: Do we need to bring this listener into the app? Supertest will attach it's own listener 💭

File: app.js [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+  getTopics
+} = require("./db/controllers/data.controller.js")
+
+const app = express();
+
+app.set('view engine', 'ejs')
+
+app.get("/api/topics", getTopics);
+
+serverInfo(app)
```
Comment (Keb84) [inline app.js]: Do you need to invoke this here?

File: db/controllers/data.controller.js [hunk 14]
```diff
@@ -0,0 +1,10 @@
+const pool = require('../connection.js')
```
Comment (Keb84) [inline db/controllers/data.controller.js]: Do you need to require your connection pool into the controller? Are you using it anywhere? 🤔

File: db/server.js [hunk 15]
```diff
@@ -0,0 +1,10 @@
+const app = require("../app.js");
+
+function serverInfo(app) {
+  app.listen(9090, (err) => {
+      if (err) console.log("Error running on server");
+      else console.log(`Server running on port 9090`);
+    });
+}
+
+module.exports = serverInfo
```
Comment (Keb84) [inline db/server.js]: Your listen file doesn't need to be exported anywhere so is this line needed?

**Review by Keb84 — CHANGES_REQUESTED** [review]
Nice work on the changes for this ticket. Please check out my comments and make the changes i've suggested then resubmit 👍

File: app.js (hunk 12 above)
Comment (jimmybombimmy) [reply app.js]: I was struggling to get this to work in any other way than this. This is annoyingly the main culprit as to why test 2 is getting redone so late. 😵

**Comment by Dutton1990** [comment]
Thanks for making the suggested changes James! You can merge and continue 👍

**Pull request merged**